File: app/analyzers/m6_domain.py
"""
M6 -- domain intelligence.

Domain age is the single most predictive cheap feature in phishing detection.
A domain registered eleven days ago asking for a wire transfer needs no ML.

Use whoisit (RDAP), not python-whois: ICANN sunset port-43 gTLD WHOIS.
Cache results in memory for this demo.

Owner: Track B
"""

from __future__ import annotations

import unicodedata
from datetime import datetime, timezone
from uuid import UUID

try:
    import whoisit
except ImportError:
    whoisit = None

try:
    import dns.resolver
except ImportError:
    dns = None

try:
    from rapidfuzz import fuzz
except ImportError:
    fuzz = None

from app.analyzers.base import register
from app.schemas.email import ParsedEmail
from app.schemas.evidence import Analyzer, Evidence

# Brands to protect (from CLAUDE.md)
PROTECTED_BRANDS = [
    "aicte-india.org",
    "sbi.co.in",
    "onlinesbi.sbi",
    "incometax.gov.in",
    "npci.org.in",
]


def _domain_age_days(domain: str) -> int | None:
    """Get domain age in days from WHOIS/RDAP."""
    if whoisit is None:
        return None

    try:
        result = whoisit.query(domain, timeout=5)

        if not result or not result.created:
            return None

        created = result.created
        if isinstance(created, list):
            created = created[0]

        if created.tzinfo is None:
            created = created.replace(tzinfo=timezone.utc)

        now = datetime.now(timezone.utc)
        age = (now - created).days
        return age
    except Exception:
        return None


def _check_dns(domain: str) -> dict:
    """Check if domain has MX and A records."""
    result = {"has_mx": False, "has_a": False, "mx_records": []}

    if dns is None:
        return result

    try:
        answers = dns.resolver.resolve(domain, "MX", lifetime=5)
        result["has_mx"] = True
        result["mx_records"] = [str(r.exchange).rstrip(".") for r in answers]
    except Exception:
        pass

    try:
        dns.resolver.resolve(domain, "A", lifetime=5)
        result["has_a"] = True
    except Exception:
        pass

    return result


# Unicode look-alikes (Cyrillic/Greek/digit) folded to their Latin twin, so a
# homograph domain reduces to the same "skeleton" as the brand it mimics.
_CONFUSABLES = str.maketrans({
    "а": "a", "е": "e", "о": "o", "р": "p", "с": "c", "х": "x", "у": "y",
    "ѕ": "s", "і": "i", "ј": "j", "н": "h", "к": "k", "м": "m", "т": "t", "в": "b",
    "ο": "o", "α": "a", "ε": "e", "ρ": "p", "τ": "t", "ν": "v", "ι": "i", "κ": "k",
    "0": "o", "1": "l", "3": "e", "4": "a", "5": "s", "7": "t",
})


def _decode_idna(domain: str) -> str:
    """Decode punycode (xn--) labels to unicode so homographs become visible."""
    if "xn--" not in domain:
        return domain
    try:
        return domain.encode("ascii").decode("idna")
    except Exception:
        return domain


def _skeleton(s: str) -> str:
    """NFKC-fold, lowercase, and map confusable characters to a Latin skeleton."""
    return unicodedata.normalize("NFKC", s).lower().translate(_CONFUSABLES)


def _tokens(skel: str) -> set[str]:
    """Alphanumeric labels/tokens (split on dot and hyphen), len>=4 only."""
    out: set[str] = set()
    for part in skel.replace("-", ".").split("."):
        if len(part) >= 4:
            out.add(part)
    return out


def _check_brand_lookalike(domain: str) -> tuple[str, str] | None:
    """
    Detect impersonation of a protected brand, including homographs and
    brand-name-in-subdomain spoofs. Returns (brand, technique) or None.

    The exact checks (homograph collision, brand-name-as-label) run even without
    rapidfuzz installed; only the fuzzy typosquat check needs it.
    """
    raw = domain.lower().strip(".")

    # A domain that IS a protected brand (or a subdomain of one) is legitimate --
    # checked on the raw string so a homograph of a brand is not let through here.
    for brand in PROTECTED_BRANDS:
        b = brand.lower()
        if raw == b or raw.endswith("." + b):
            return None

    uni = _decode_idna(raw)
    skel = _skeleton(uni)
    labels = skel.split(".")
    non_tld = labels[:-1] if len(labels) > 1 else labels   # exclude the TLD label
    non_tld_tokens = _tokens(".".join(non_tld))
    sig = max(non_tld, key=len) if non_tld else ""          # most significant label
    homographed = (uni != raw) or ("xn--" in raw) or any(ord(c) > 127 for c in raw)

    for brand in PROTECTED_BRANDS:
        b = brand.lower()
        b_skel = _skeleton(b)
        b_core = b_skel.split(".")[0]           # e.g. "sbi", "aicte-india", "npci"

        # 1. Whole-domain homograph collision (visually identical after folding).
        if skel == b_skel:
            return (brand, "homograph" if homographed else "typosquat")

        # 2. The brand's core label appears verbatim outside the TLD position
        #    (dotted label or hyphen token) -> subdomain/prefix spoof, e.g.
        #    sbi.co.in.secure-login.tk or npci-refund.tk.
        if b_core in non_tld or b_core in non_tld_tokens:
            return (brand, "brand-name-in-domain")

        # 3. Fuzzy typosquat of the significant label (paypa1 -> paypal).
        if fuzz is not None and len(b_core) >= 4:
            if fuzz.ratio(sig, b_core) >= 82:
                return (brand, "homograph" if homographed else "typosquat")
            for bt in (_tokens(b_core) or {b_core}):
                if len(bt) >= 5 and any(fuzz.ratio(t, bt) >= 85 for t in non_tld_tokens):
                    return (brand, "typosquat")

    return None


@register(Analyzer.M6_DOMAIN)
async def analyze(case_id: UUID, email: ParsedEmail) -> list[Evidence]:
    """Analyze sender domain for age, DNS records, and brand lookalike."""
    ev: list[Evidence] = []

    from_domain = email.from_addr.split("@")[-1].lower() if email.from_addr else None
    if not from_domain:
        return [
            Evidence.unavailable(
                case_id,
                Analyzer.M6_DOMAIN,
                "domain_age_lt_30d",
                "No From domain found",
            )
        ]

    # Check domain age
    age_days = _domain_age_days(from_domain)
    if age_days is not None and age_days < 30:
        ev.append(
            Evidence.triggered(
                case_id,
                Analyzer.M6_DOMAIN,
                "domain_age_lt_30d",
                detail={
                    "age_days": age_days,
                    "explanation": f"Domain registered only {age_days} days ago — common phishing pattern.",
                },
            )
        )

    # Check DNS records
    dns_result = _check_dns(from_domain)

    if not dns_result["has_mx"]:
        ev.append(
            Evidence.triggered(
                case_id,
                Analyzer.M6_DOMAIN,
                "domain_no_mx",
                detail={
                    "domain": from_domain,
                    "explanation": "No MX (mail server) records found — domain is not properly configured to send email.",
                },
            )
        )

    if not dns_result["has_a"]:
        ev.append(
            Evidence.triggered(
                case_id,
                Analyzer.M6_DOMAIN,
                "domain_does_not_resolve",
                detail={
                    "domain": from_domain,
                    "explanation": "Domain does not resolve (no A record) — possibly abandoned or spoofed.",
                },
            )
        )

    # Check for brand lookalike (homograph / typosquat / subdomain spoof)
    lookalike = _check_brand_lookalike(from_domain)
    if lookalike:
        brand, technique = lookalike
        reasons = {
            "homograph": f"uses look-alike characters to imitate '{brand}'",
            "typosquat": f"is a near-misspelling of protected brand '{brand}'",
            "brand-name-in-domain": f"places the protected brand '{brand}' in a domain it does not own",
        }
        ev.append(
            Evidence.triggered(
                case_id,
                Analyzer.M6_DOMAIN,
                "brand_lookalike_domain",
                detail={
                    "domain": from_domain,
                    "similar_to": brand,
                    "technique": technique,
                    "explanation": f"Domain '{from_domain}' {reasons.get(technique, 'imitates a protected brand')}.",
                },
            )
        )

    # If nothing found, return CLEAR
    if not ev:
        ev.append(
            Evidence.clear(
                case_id,
                Analyzer.M6_DOMAIN,
                "domain_age_lt_30d",
                detail={
                    "domain": from_domain,
                    "explanation": "Domain appears legitimate (not new, has MX records, resolves, not a lookalike).",
                },
            )
        )

    return ev

File: web/src/lib/analyzers/m6_domain.ts
/**
 * M6 -- domain intelligence (ported from app/analyzers/m6_domain.py).
 *
 * Domain age is the single most predictive cheap feature in phishing. Uses RDAP
 * (over HTTPS) not port-43 WHOIS. Brand-lookalike detection folds Unicode
 * confusables and punycode to a Latin "skeleton" so homographs collapse onto the
 * brand they mimic; the fuzzy typosquat check uses a normalized-Indel ratio
 * equivalent to rapidfuzz.fuzz.ratio.
 */

import { resolveMx, resolve4 } from 'node:dns/promises';
import { domainToUnicode } from 'node:url';
import { Analyzer, clear, triggered, unavailable, type Evidence } from '../schemas/evidence';
import type { ParsedEmail } from '../schemas/email';
import { register } from './base';

const PROTECTED_BRANDS = [
  'aicte-india.org',
  'sbi.co.in',
  'onlinesbi.sbi',
  'incometax.gov.in',
  'npci.org.in',
];

// Unicode look-alikes folded to their Latin twin.
const CONFUSABLES: Record<string, string> = {
  а: 'a', е: 'e', о: 'o', р: 'p', с: 'c', х: 'x', у: 'y',
  ѕ: 's', і: 'i', ј: 'j', н: 'h', к: 'k', м: 'm', т: 't', в: 'b',
  ο: 'o', α: 'a', ε: 'e', ρ: 'p', τ: 't', ν: 'v', ι: 'i', κ: 'k',
  '0': 'o', '1': 'l', '3': 'e', '4': 'a', '5': 's', '7': 't',
};

function fold(s: string): string {
  let out = '';
  for (const ch of s) out += CONFUSABLES[ch] ?? ch;
  return out;
}

function decodeIdna(domain: string): string {
  if (!domain.includes('xn--')) return domain;
  try {
    return domainToUnicode(domain) || domain;
  } catch {
    return domain;
  }
}

function skeleton(s: string): string {
  return fold(s.normalize('NFKC').toLowerCase());
}

function tokens(skel: string): Set<string> {
  const out = new Set<string>();
  for (const part of skel.replace(/-/g, '.').split('.')) {
    if (part.length >= 4) out.add(part);
  }
  return out;
}

/** rapidfuzz.fuzz.ratio == normalized Indel similarity == 200*LCS/(len1+len2). */
export function fuzzRatio(a: string, b: string): number {
  if (!a.length && !b.length) return 100;
  const m = a.length;
  const n = b.length;
  const dp = new Array(n + 1).fill(0);
  for (let i = 1; i <= m; i++) {
    let prev = 0;
    for (let j = 1; j <= n; j++) {
      const tmp = dp[j];
      dp[j] = a[i - 1] === b[j - 1] ? prev + 1 : Math.max(dp[j], dp[j - 1]);
      prev = tmp;
    }
  }
  const lcs = dp[n];
  return (200 * lcs) / (m + n);
}

export function checkBrandLookalike(domain: string): [string, string] | null {
  const raw = domain.toLowerCase().replace(/^\.+|\.+$/g, '');

  for (const brand of PROTECTED_BRANDS) {
    const b = brand.toLowerCase();
    if (raw === b || raw.endsWith('.' + b)) return null; // is (a subdomain of) the real brand
  }

  const uni = decodeIdna(raw);
  const skel = skeleton(uni);
  const labels = skel.split('.');
  const nonTld = labels.length > 1 ? labels.slice(0, -1) : labels;
  const nonTldTokens = tokens(nonTld.join('.'));
  // Exact hyphen/underscore/dot-delimited parts, NOT length-filtered -- so a
  // short brand core like "sbi" (3 chars, dropped by tokens()) is still caught
  // as an exact standalone token, e.g. sbi-secure-login.com.
  const nonTldParts = new Set(nonTld.join('.').replace(/[-_]/g, '.').split('.').filter(Boolean));
  const sig = nonTld.reduce((a, b) => (b.length > a.length ? b : a), '');
  const homographed = uni !== raw || raw.includes('xn--') || [...raw].some((c) => c.charCodeAt(0) > 127);

  for (const brand of PROTECTED_BRANDS) {
    const b = brand.toLowerCase();
    const bSkel = skeleton(b);
    const bCore = bSkel.split('.')[0];

    if (skel === bSkel) return [brand, homographed ? 'homograph' : 'typosquat'];

    if (nonTld.includes(bCore) || nonTldTokens.has(bCore) || (bCore.length >= 3 && nonTldParts.has(bCore))) {
      return [brand, 'brand-name-in-domain'];
    }

    if (bCore.length >= 4) {
      if (fuzzRatio(sig, bCore) >= 82) return [brand, homographed ? 'homograph' : 'typosquat'];
      const brandTokens = tokens(bCore).size ? tokens(bCore) : new Set([bCore]);
      for (const bt of brandTokens) {
        if (bt.length >= 5 && [...nonTldTokens].some((t) => fuzzRatio(t, bt) >= 85)) {
          return [brand, 'typosquat'];
        }
      }
    }
  }
  return null;
}

const DNS_NET_ERRORS = new Set(['EAI_AGAIN', 'ETIMEOUT', 'ETIMEDOUT', 'ESERVFAIL', 'ECONNREFUSED', 'ECONNRESET', 'ENETUNREACH', 'ENOTIMP']);
function isDnsNetworkError(e: unknown): boolean {
  const code = (e as { code?: string })?.code;
  return code ? DNS_NET_ERRORS.has(code) : false;
}

// --- RDAP domain age ---------------------------------------------------------
// `unavailable` = we could not reach RDAP (offline / timeout / server error),
// so age is unknown and MUST surface as UNAVAILABLE, not as a silent "not new".
// A definitive 404 (domain not in RDAP / unsupported TLD) is `ageDays: null`
// WITHOUT unavailable — age simply unknown, no network fault.
export interface DomainAge { ageDays: number | null; unavailable: boolean }

export async function domainAgeDays(domain: string, timeoutMs = 4000): Promise<DomainAge> {
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    const res = await fetch(`https://rdap.org/domain/${encodeURIComponent(domain)}`, {
      signal: ctrl.signal,
      headers: { accept: 'application/rdap+json' },
    });
    if (res.status >= 500) return { ageDays: null, unavailable: true }; // RDAP server error
    if (!res.ok) return { ageDays: null, unavailable: false };          // 404 etc: age simply unknown
    const data = (await res.json()) as { events?: Array<{ eventAction?: string; eventDate?: string }> };
    const reg = (data.events ?? []).find((e) => e.eventAction === 'registration');
    if (!reg?.eventDate) return { ageDays: null, unavailable: false };
    const created = new Date(reg.eventDate);
    if (Number.isNaN(created.getTime())) return { ageDays: null, unavailable: false };
    return { ageDays: Math.floor((Date.now() - created.getTime()) / 86_400_000), unavailable: false };
  } catch {
    return { ageDays: null, unavailable: true }; // network error / abort (timeout): unreachable
  } finally {
    clearTimeout(timer);
  }
}

// `dnsDown` = the resolver itself was unreachable (offline / transient), which is
// NOT the same as a domain that genuinely has no A/MX records.
async function checkDns(domain: string): Promise<{ hasMx: boolean; hasA: boolean; dnsDown: boolean }> {
  let hasMx = false;
  let hasA = false;
  let dnsDown = false;
  try {
    hasMx = (await resolveMx(domain)).length > 0;
  } catch (e) {
    if (isDnsNetworkError(e)) dnsDown = true; // else: genuinely no MX
  }
  try {
    hasA = (await resolve4(domain)).length > 0;
  } catch (e) {
    if (isDnsNetworkError(e)) dnsDown = true; // else: genuinely no A
  }
  return { hasMx, hasA, dnsDown };
}

export async function analyze(caseId: string, email: ParsedEmail): Promise<Evidence[]> {
  const ev: Evidence[] = [];
  const fromDomain = email.from_addr ? email.from_addr.split('@').pop()!.toLowerCase() : null;
  if (!fromDomain) {
    return [unavailable(caseId, Analyzer.M6_DOMAIN, 'domain_age_lt_30d', 'No From domain found')];
  }

  const { ageDays, unavailable: ageDown } = await domainAgeDays(fromDomain);
  if (ageDown) {
    ev.push(unavailable(caseId, Analyzer.M6_DOMAIN, 'domain_age_lt_30d',
      `Registration age for ${fromDomain} could not be retrieved (RDAP unreachable) — a newly-registered domain cannot be ruled out.`));
  } else if (ageDays !== null && ageDays < 30) {
    ev.push(
      triggered(caseId, Analyzer.M6_DOMAIN, 'domain_age_lt_30d', {
        age_days: ageDays,
        explanation: `Domain registered only ${ageDays} days ago — common phishing pattern.`,
      }),
    );
  }

  const { hasMx, hasA, dnsDown } = await checkDns(fromDomain);
  if (dnsDown) {
    // Resolver unreachable (offline / transient): do NOT fabricate no-MX /
    // does-not-resolve for every domain (that was flagging even the benign
    // control offline). Surface it as UNAVAILABLE, not a definite finding.
    ev.push(unavailable(caseId, Analyzer.M6_DOMAIN, 'domain_does_not_resolve',
      `DNS for ${fromDomain} could not be resolved (resolver unreachable) — resolve/MX state unknown.`));
  } else {
    if (!hasMx) {
      ev.push(
        triggered(caseId, Analyzer.M6_DOMAIN, 'domain_no_mx', {
          domain: fromDomain,
          explanation: 'No MX (mail server) records found — domain is not configured to receive email.',
        }),
      );
    }
    if (!hasA) {
      ev.push(
        triggered(caseId, Analyzer.M6_DOMAIN, 'domain_does_not_resolve', {
          domain: fromDomain,
          explanation: 'Domain does not resolve (no A record) — possibly abandoned or spoofed.',
        }),
      );
    }
  }

  const lookalike = checkBrandLookalike(fromDomain);
  if (lookalike) {
    const [brand, technique] = lookalike;
    const reasons: Record<string, string> = {
      homograph: `uses look-alike characters to imitate '${brand}'`,
      typosquat: `is a near-misspelling of protected brand '${brand}'`,
      'brand-name-in-domain': `places the protected brand '${brand}' in a domain it does not own`,
    };
    ev.push(
      triggered(caseId, Analyzer.M6_DOMAIN, 'brand_lookalike_domain', {
        domain: fromDomain,
        similar_to: brand,
        technique,
        explanation: `Domain '${fromDomain}' ${reasons[technique] ?? 'imitates a protected brand'}.`,
      }),
    );
  }

  if (!ev.length) {
    // Reached only when nothing triggered AND nothing was unavailable, i.e. the
    // checks actually ran. Don't assert "not new" unless age was determined.
    const ageNote = ageDays !== null ? `${ageDays}d old` : 'age not in RDAP';
    ev.push(
      clear(caseId, Analyzer.M6_DOMAIN, 'domain_age_lt_30d', {
        domain: fromDomain,
        explanation: `Domain resolves with a valid MX record and is not a brand lookalike (${ageNote}).`,
      }),
    );
  }

  return ev;
}

register(Analyzer.M6_DOMAIN, analyze);
